### PR TITLE
Set images as attachments of posts

### LINKED ISSUE #4942: Inserted images are not set as attachments or have dimensions
When inserting an image into a post, the image is not properly attached to the post so meta data can be accessed. The image also is set without any dimensions in the HTML tag. This is causing a problem in the browser, where images cannot be displayed in the posts list for a site in Calypso.

See this issue for more details: https://github.com/Automattic/wp-calypso/issues/3954

### Steps to reproduce the behavior

Create a new post in the editor and upload a single image. Publish the post. Go to Calypso in the browser and look at your /posts page for the site. You will notice the image does not show for the post in the list. This is because there is no attachment data to fetch the size of the image, and there's no dimensions set on the image.

##### Tested on [device], iOS [version]

Latest beta, and stable. iOS 9.3 public beta latest, iPhone 6 Plus

### PR DESCRIPTION
**Pull request by SergioEstevao**
Fixes #4942 

To test:

Scenario 1:
 - Start a new post
 - Add a new device media to it (video or image)
 - Publish the post
 - Check on the server if the new media is attached to the new post

Scenario 2:
 - Start a new post
 - Add some content
 - Save the post or Publish
 - Open the post again 
 - Add a new device media to it (video or image)
 - Save the post
 - Check on the server if the new media is attached to the new post

Check on: WordPress.com blogs, jetpack connect blogs and self-hosted blogs,use the visual editor and the legacy editor.

Needs review: @koke

### CODE REVIEW

File: WordPress/Classes/Networking/MediaServiceRemoteREST.m [hunk 1]
```diff
@@ -122,10 +122,13 @@ - (void)createMedia:(RemoteMedia *)media
     NSString *apiPath = [NSString stringWithFormat:@"sites/%@/media/new", self.siteID];
     NSString *requestUrl = [self pathForEndpoint:apiPath
                                      withVersion:ServiceRemoteRESTApiVersion_1_1];
-    
+    NSMutableDictionary *parameters = [NSMutableDictionary dictionaryWithDictionary:@{}];
+    if ([media.postID compare:@(0)] == NSOrderedDescending) {
+        parameters[@"attrs[0][parent_id]"] = media.postID;
```
Comment (koke) [inline WordPress/Classes/Networking/MediaServiceRemoteREST.m]: Do you know if this works if `postID` is `nil`?

File: WordPress/Classes/Networking/MediaServiceRemote.h [hunk 2]
```diff
@@ -15,6 +15,17 @@
             failure:(void (^)(NSError *error))failure;
 
 /**
+ *  Update media details on the server
+ *
+ *  @param media   the media object to update
+ *  @param success uccess a block to be executed when the request finishes with success.
```
Comment (koke) [inline WordPress/Classes/Networking/MediaServiceRemote.h]: Small typo: `uccess`

File: WordPress/Classes/Networking/MediaServiceRemoteREST.m (hunk 1 above)
Comment (koke) [inline WordPress/Classes/Networking/MediaServiceRemoteREST.m]: Doh, I just saw the `if` above. It's not very obvious with the `compare:`

File: WordPress/Classes/Services/MediaService.m [hunk 3]
```diff
@@ -186,6 +187,54 @@ - (void)uploadMedia:(Media *)media
...
 }
 
+- (void)updateMedia:(Media *)media
+            success:(void (^)())success
+            failure:(void (^)(NSError *error))failure
+{
+    id<MediaServiceRemote> remote = [self remoteForBlog:media.blog];
+    RemoteMedia *remoteMedia = [self remoteMediaFromMedia:media];
+
+    [[ContextManager sharedInstance] saveContext:self.managedObjectContext];
```
Comment (koke) [inline WordPress/Classes/Services/MediaService.m]: I'm curious: why do we need to save here? And if we do, shouldn't we wait for the save to finish?

File: WordPress/Classes/Services/MediaService.m [hunk 4]
```diff
@@ -186,6 +187,54 @@ - (void)uploadMedia:(Media *)media
...
+                    success();
+                }
+            }];
+        }];
+    };
+    void (^failureBlock)(NSError *error) = ^(NSError *error) {
+        [self.managedObjectContext performBlock:^{
+            Media *mediaInContext = (Media *)[self.managedObjectContext existingObjectWithID:mediaObjectID error:nil];
+            if (mediaInContext) {
+                mediaInContext.remoteStatus = MediaRemoteStatusFailed;
```
Comment (koke) [inline WordPress/Classes/Services/MediaService.m]: Wouldn't this leave the media as if it had failed uploading? Would it then show a retry button to upload again?

**Comment by koke** [comment]
Other than those questions, this is looking good :smile:

File: WordPress/Classes/Services/MediaService.m (hunk 4 above)
Comment (SergioEstevao) [inline WordPress/Classes/Services/MediaService.m]: You are right need to remove this status changing from here and above.

**Comment by SergioEstevao** [comment]
@koke ready for another round.

**Comment by koke** [comment]
:shipit:

**Pull request merged**